src/RealFourierArray.jl

```diff
@@ -0,0 +1,91 @@
+# Lazy structure for an array whose real and fourier part can be accessed by A.real and A.fourier
+# For normalization conventions, see PlaneWaveBasis.jl
+
+import Base.getproperty, Base.setproperty!, Base.fieldnames
+import Base.*, Base.+, Base./
+import Base.eltype
+
+## these functions are minimally optimized for the moment
+
+## TODO this always allocates both real and fourier parts, which might be a bit wasteful
+## TODO implement broadcasting. This seems to be non-trivial, so I just implement a few simple methods
+## TODO implement compressed representation in Fourier space of real arrays
+
+# Tr is the underlying real type
+# T is the type of the array in real space
+# TODO make this concrete by parametrizing on the type of the arrays
+mutable struct RealFourierArray{Tr <: Real, T <: Union{Tr, Complex{Tr}}}
+    basis::PlaneWaveBasis{Tr}
+    _real::AbstractArray{T, 3}
+    _fourier::AbstractArray{Complex{Tr}, 3}
+    _state::Int # 0: only _real set. 1: only _fourier set. 2: both set
+end
+# Type of the real part
+Base.eltype(A::RealFourierArray{Tr, T}) where {Tr, T} = T
+
+# Constructors
+RealFourierArray(basis, real, fourier) = RealFourierArray(basis, real, fourier, 2)
+function RealFourierArray(basis::PlaneWaveBasis{T}; iscomplex=false) where {T}
+    from_real(basis, zeros(iscomplex ? complex(T) : T, basis.fft_size...))
+end
+
+function from_real(basis, real_part::AbstractArray{T}) where {T <: Number}
+    RealFourierArray{real(T), T}(basis, real_part, similar(real_part, complex(T)), 0)
+end
+# this will necessarily be a real array
+function from_fourier(basis, fourier_part::AbstractArray{T}; isreal=false) where {T <: Complex}
```

Comment:
To what does this comment refer to? I'd make an empty line to make that clear.

Reply:
Also `isreal` is not that clear to me. What you want to say is whether the elements are purely real in the real-space representation, right. How about using a Symbol flag `:purely_real`  for that? Then one could (in theory) extend to `:purely_imaginary` and so on.

Reply:
The comment was obsolete. The issreal seems fine to me, no? See previous comment. There's `Array{Treal}` and `Array{Complex{Treal}}`, it would be weird to have a type just for an imaginary array

Reply:
I think `isreal` is at least misleading, because you it could ring the bell of "what I actually pass is the real-space representation" or "what I pass has only real-valued entries",  but then the function is named `from_fourier` and takes complex arrays. I just think reading
```julia
from_fourier(basis, array, isreal=true)
```
versus something like
```julia
from_fourier(basis, array, :assume_real)
```
or
```julia
from_fourier(basis, array, assume_real=true)
```
where the latter two are clearer in my opinion.